c_emulator/rvfi_dii.cpp

```diff
@@ -14,8 +14,152 @@
+      .rvfi_cmd = (uint8_t)((value >> 48) & 0xFF),
+      .padding = (uint8_t)((value >> 56) & 0xFF),
+  };
+}
+
+RVFI_DII_Execution_Packet_V1
+RVFI_DII_Execution_Packet_V1::rvfi_get_v2_support_packet()
+{
+  // Returning 0x3 (using the unused high bits) in halt instead of 0x1 means
+  // that we support the version 2 wire format. This is required to keep
+  // backwards compatibility with old implementations that do not support
+  // the new trace format.
+  //  res = {};
+  return RVFI_DII_Execution_Packet_V1 {
+      .rvfi_halt = 0x03,
+  };
+}
+
+// ****************************************************************************
+// RVFI Handler Implementation
+
+RVFI_DII_Execution_Packet_V1 rvfi_handler::rvfi_get_exec_packet_v1()
+{
+  return RVFI_DII_Execution_Packet_V1 {
+      .rvfi_order = rvfi_inst_data.rvfi_order,
+      .rvfi_pc_rdata = rvfi_pc_data.rvfi_pc_rdata,
+      .rvfi_pc_wdata = rvfi_pc_data.rvfi_pc_wdata,
+      .rvfi_insn = rvfi_inst_data.rvfi_insn,
+
+      .rvfi_rs1_data = rvfi_int_data.rvfi_rs1_rdata,
+      .rvfi_rs2_data = rvfi_int_data.rvfi_rs2_rdata,
+      .rvfi_rd_wdata = rvfi_int_data.rvfi_rd_wdata,
+
+      .rvfi_mem_addr = rvfi_mem_data.rvfi_mem_addr,
+      .rvfi_mem_rdata = rvfi_mem_data.rvfi_mem_rdata[0],
+      .rvfi_mem_wdata = rvfi_mem_data.rvfi_mem_wdata[0],
+      .rvfi_mem_rmask = uint8_t(rvfi_mem_data.rvfi_mem_rmask),
+      .rvfi_mem_wmask = uint8_t(rvfi_mem_data.rvfi_mem_wmask),
+
+      .rvfi_rs1_addr = rvfi_int_data.rvfi_rs1_addr,
+      .rvfi_rs2_addr = rvfi_int_data.rvfi_rs2_addr,
+      .rvfi_rd_addr = rvfi_int_data.rvfi_rd_addr,
+
+      .rvfi_trap = rvfi_inst_data.rvfi_trap,
+      .rvfi_halt = rvfi_inst_data.rvfi_halt,
+      .rvfi_intr = rvfi_inst_data.rvfi_intr,
+  };
+}
+
+uint64_t rvfi_handler::rvfi_get_v2_trace_size()
+{
+  uint64_t trace_size = sizeof(RVFI_DII_Execution_Packet_V2);
+  if (rvfi_int_data_present) {
+    trace_size += sizeof(RVFI_DII_Execution_Packet_Ext_Integer);
+  }
+  if (rvfi_mem_data_present) {
+    trace_size += sizeof(RVFI_DII_Execution_Packet_Ext_MemAccess);
+  }
+  return trace_size;
+}
+
+RVFI_DII_Execution_Packet_V2 rvfi_handler::rvfi_get_exec_packet_v2()
+{
+  uint64_t available_fields = 0;
+  if (rvfi_int_data_present) {
+    available_fields |= RVFI_INTEGER_DATA;
+  }
+  if (rvfi_mem_data_present) {
+    available_fields |= RVFI_MEM_DATA;
+  }
+  return RVFI_DII_Execution_Packet_V2 {
+      .trace_size = rvfi_get_v2_trace_size(),
+      .basic_data = rvfi_inst_data,
+      .pc_data = rvfi_pc_data,
+      .available_fields = available_fields,
+  };
+}
+
+void rvfi_handler::rvfi_zero_exec_packet()
+{
+  rvfi_inst_data = {};
+  rvfi_pc_data = {};
+  rvfi_int_data = {};
+  rvfi_mem_data = {};
+  rvfi_int_data_present = false;
+  rvfi_mem_data_present = false;
+}
+
+void rvfi_handler::rvfi_halt_exec_packet()
+{
+  rvfi_inst_data.rvfi_halt = 0x01;
+}
+
+void rvfi_handler::print_rvfi_exec()
+{
+  print_sbits("rvfi_intr     : ", make_sbits(rvfi_inst_data.rvfi_intr));
```

Comment:
```suggestion
  print_sbits("rvfi_intr     : %#016jx\n", (uintmax_t)rvfi_inst_data.rvfi_intr);
```

How about this instead? Uintmax_t and %j printf modifier to deal with different uint64_t representations. Or you could use the PRIx macros

Reply:
`print_sbits` currently decides whether to print in binary or hex based on the `sbits.len`, and it will only print the length of `sbits.len`. (The code is mainly modified from `print_bits` which is used for mpz_xxx.)

Using `uintmax_t` is obviously not very good?

Reply:
But all the callers here have a bit width that is a multiple of 8, so always print in hex. Printf with the PRI macros seems much easier